### PR TITLE
Simplify capabilities (Task #4194)

### PR DESCRIPTION
**Pull request by nicosp**
Major change.

- Converts the plugin from an event listener to per-table behavior.
- A SubjectInterface is introduced to clearly define the contract between application and plugin.
- Middleware for easiy integration to authentication middleware.
- Entity-based Authorization adapter for AuthComponent (optional).

### CODE REVIEW

File: src/EntityAccess/AccessControlTrait.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+     * It uses isTableActionAuthorized for authorization.
+     *
+     * @param Controller $controller The controller to check
+     * @param string $action The action to check
+     * @param ?string $entityId The entity to check
+     * @param ?SubjectInterface $user The user
+     *
+     * @return bool
+     */
+    public function isActionAuthorized(Controller $controller, string $action, ?string $entityId, ?SubjectInterface $user = null): bool
```
Comment (softius) [inline src/EntityAccess/AccessControlTrait.php:22]: @nicosp shall we rename this to `isControllerActionAuthorized` ?

File: src/EntityAccess/CapabilitiesUtil.php [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+     * It works by scanning composer's autoload classmap
+     *
+     * @return string[]
+     */
+    public static function getTables(): array
+    {
+        $map = require 'vendor/composer/autoload_classmap.php';
+
+        $tables = [];
+        foreach ($map as $class => $path) {
```
Comment (softius) [inline src/EntityAccess/CapabilitiesUtil.php:27]: @nicosp there are a few app helpers that we could possibly use here like `App::shortName`

File: src/EntityAccess/CapabilitiesUtil.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    /**
+     * Gets all capabilities defined in the model
+     *
+     * @return mixed[]
+     */
+    public static function getModelStaticCapabities(Table $table): array
+    {
+        $tableCapabilities = [];
+
+        if ($table->hasBehavior('Authorized')) {
```
Comment (softius) [inline src/EntityAccess/CapabilitiesUtil.php:69]: @nicosp one more return early.

File: src/EntityAccess/CapabilitiesUtil.php [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    /**
+     * Gets all capabilities defined in the model
+     *
+     * @return mixed[]
+     */
+    public static function getModelCapabilityAssociations(Table $table): array
+    {
+        $tableCapabilities = [];
+
+        if ($table->hasBehavior('Authorized')) {
```
Comment (softius) [inline src/EntityAccess/CapabilitiesUtil.php:88]: @nicosp same here.

File: src/EntityAccess/EntityCapabilityRule.php [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        $entity = $query->where($exp)->first();
+
+        return $entity !== null;
+    }
+
+    /** Create unique alias for the table
+     *
+     * @return string unique alias for this table
+     */
+    private function aliasTable(): string
```
Comment (softius) [inline src/EntityAccess/EntityCapabilityRule.php:118]: @nicosp lets explain why we needed this for future reference.

File: src/EntityAccess/EntityCapabilityRule.php [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            $driver = $this->table->getConnection()->getDriver();
+            $quotedAlias = $driver->quoteIdentifier($this->aliasTable());
+            $quotedSubject = $driver->quote($this->subject->getId(), \PDO::PARAM_STR);
+
+            /*
+             * Create inner query. We cannot use parameters since we are creating
+             * an sql string.
+             */
+            $innerQuery = $this->table->query()
+            ->applyOptions(['filterQuery' => true ])
```
Comment (softius) [inline src/EntityAccess/EntityCapabilityRule.php:182]: @nicosp something is off with the spacing here.

File: src/EntityAccess/Event/ModelInitializeListener.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        $key = str_replace('.', '_', $table->getRegistryAlias());
+
+        $config = Configure::read('RolesCapabilities.tables.' . $key);
+        if ($config === null) {
+            $config = Configure::read('RolesCapabilities.tables.*');
+        }
+
+        if ($config !== null) {
```
Comment (softius) [inline src/EntityAccess/Event/ModelInitializeListener.php:47]: @nicosp perhaps you can add a return in above `if` and remove the extra if here.

**Review by softius — CHANGES_REQUESTED** [review]

File: src/EntityAccess/AccessControlTrait.php (hunk 1 above)
Comment (nicosp) [reply src/EntityAccess/AccessControlTrait.php]: Done

File: src/EntityAccess/CapabilitiesUtil.php (hunk 3 above)
Comment (nicosp) [reply src/EntityAccess/CapabilitiesUtil.php]: Done

File: src/EntityAccess/EntityCapabilityRule.php (hunk 5 above)
Comment (nicosp) [reply src/EntityAccess/EntityCapabilityRule.php:121]: Done

File: src/EntityAccess/Event/ModelInitializeListener.php (hunk 7 above)
Comment (nicosp) [reply src/EntityAccess/Event/ModelInitializeListener.php]: Done

File: src/EntityAccess/CapabilitiesUtil.php (hunk 2 above)
Comment (nicosp) [reply src/EntityAccess/CapabilitiesUtil.php:28]: Done

File: src/EntityAccess/EntityCapabilityRule.php [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+     * @inheritdoc
+     */
+    public function expression(Query $query): QueryExpression
+    {
+        $capabilities = $this->getCapabilities();
+
+        $exp = $query->newExpr();
+        $exp->setConjunction('OR');
+
+        $isEmpty = true;
```
Comment (softius) [inline src/EntityAccess/EntityCapabilityRule.php:136]: @nicosp i dont follow why we need `$isEmpty` here can you please advise and perhaps add a comment?

**Review by softius — CHANGES_REQUESTED** [review]

**Review by softius — APPROVED** [review]

**Pull request closed without merging**